073-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,139 @@
+Including the state in the item name facilitates inspection of the Secret.
+
+The existing Secrets, named using `${cluster}-${ca}-certs` and passed via Secret volume mounts to containers would not change under this proposal.
+
+Problems from multiple processes concurrently updating the secret are avoided by using `metadata.resourceVersion` to make conditional requests when updating that Secret. Note also that only one operator makes each state transition.
+
+### CA state machine
+
+The "CA certificate trust state machine" has the following states:
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_UNUSED // trusted everywhere it needs to be, but no EE certificates yet issued
+   |
+   v
+  TRUSTED_USED // trusted everywhere, EE certificates issued
+   |
+   v
+  PHASE_OUT // trusted everywhere, but being phased out
+   |
+   v
+  ZOMBIE // still trusted by some component, but not relied on by any party.
+```
+
+New cluster creation process:
+1. The CaReconciler creates the CA certificates and places them in the relevant Secret with the state UNTRUSTED
+2. The cluster and user operator, during their respective reconcile loops:
+   1. Add the CA certificate to the relevant component truststore secret and change the state to TRUSTED_UNUSED
+   2. Generate end entity certificates using the CA and add the EE certificate to the relevant component keystore secret and change the state to TRUSTED_USED
+
+Update process (e.g. CA key has been updated):
+1. The CaReconciler generates a new CA certificate and places it in the relevant Secret with the state UNTRUSTED. It also updates the old CA certificate to have the state PHASE_OUT.
+2. The cluster and user operator, during their respective reconcile loop:
+   1. Add the CA certificate to the relevant component truststore secret and change the state to TRUSTED_UNUSED
+   2. Generate end entity certificates using the CA and add the EE certificate to the relevant component keystore secret and change the state to TRUSTED_USED
+   3. Update the old CA certificate to ZOMBIE, indicating that no EE certicates are using the CA certificate
+3. In a future loop the CaReconciler removes the CA certificate from the shared Secret
+4. The cluster and user operator remove the CA certificate from the truststore Secret
+
+Notes:
+1. The key for the self-signed cluster CA in a Kafka cluster called **foo**, can be stored in `foo-cluster-ca`.
+2. The strimzi.io/cluster-ca-key-generation can be used by the cluster operator when managing the self-signed cluster CA
+3. The strimzi.io/cluster-ca-cert-generation on components can be used to store the fingerprint of the issuing CA, rather than an increasing integer
+
+### Future changes
+
+Once this proposal is implemented, it allows in future to update to change both the issuing of end entity certificates and the management of the CAs to be done by certificate management solutions, similarly to the [closed proposal][pr46].
```

Comment:
I do think it's worth adding a sentence or two about the overall direction/plan. What we want to achieve, and what this Proposal takes a step towards, is having each `Secret` have a clear, well-defined role. Other steps required are:

1. Changing the EE cert issuance process such that EE certificates are issued, their issuer inspected and if not already trusted then the new EE certificates are held in a "staging Secret" while trust in the CA certificate is rolled out. We can do this refactoring without adding support for Cert Manager.
2. Having each broker have its own Secret for its EE certificate and key. This isn't strictly required for Cert Manager support, but it does enable larger clusters.
3. Add support for Cert Manager.

Reply:
I've added some more information, hopefully i's clearer now